src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+
+  /**
+   *
+   */
+  constructor(message: string, jobId: string) {
+    super(message);
+    this.name = 'JobPollingTimeout';
+    this.jobId = jobId;
+  }
+}
+
+class BulkApiV2<S extends Schema> extends HttpApi<S> {
+  hasErrorInResponseBody(body: any) {
+    return (
+      Array.isArray(body) &&
+      typeof body[0] === 'object' &&
+      'errorCode' in body[0]
+    );
+  }
+
+  isSessionExpired(response: HttpResponse): boolean {
+    return (
+      response.statusCode === 401 && /INVALID_SESSION_ID/.test(response.body)
+    );
+  }
+
+  parseError(body: any) {
+    return {
+      errorCode: body[0].errorCode,
+      message: body[0].message,
+    };
+  }
+}
+
+export class BulkV2<S extends Schema> {
+  private connection: Connection<S>;
+  private logger: Logger;
+
+  /**
+   * Polling interval in milliseconds
+   */
+  pollInterval = 1000;
+
+  /**
+   * Polling timeout in milliseconds
+   */
+  pollTimeout = 10000;
+
+  constructor(connection: Connection<S>) {
+    this.connection = connection;
+    this.logger = connection._logger;
+    this.logger = this.connection._logLevel
+      ? getLogger('bulk2').createInstance(this.connection._logLevel)
+      : getLogger('bulk2');
+  }
+
+  /**
+   * Create an instance of an ingest job object.
+   *
+   * @params {NewIngestJobOptions} options object
+   * @returns {IngestJobV2} An ingest job instance
+   * @example
+   * // Upsert records to the Account object.
+   *
+   * const job = connection.bulk2.createJob({
+   *   operation: 'insert'
+   *   object: 'Account',
+   * });
+   *
+   * // create the job in the org
+   * await job.open()
+   *
+   * // upload data
+   * await job.uploadData(csvFile)
+   *
+   * // finished uploading data, mark it as ready for processing
+   * await job.close()
+   */
+  public createJob<Opr extends IngestOperation>(
+    options: NewIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Get an ingest job instance specified by a given job ID
+   *
+   * @param options Options object with a job ID
+   * @returns IngestJobV2 An ingest job
+   */
+  public job<Opr extends IngestOperation>(
+    options: ExistingIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Create, upload, and start bulkload job
+   */
+  async loadAndWaitForResults(
+    options: NewIngestJobOptions &
+      Partial<BulkV2PollingOptions> & {
+        input: Record[] | Readable | string;
+      },
+  ): Promise<IngestJobV2Results<S>> {
+    if (!options.pollTimeout) options.pollTimeout = this.pollTimeout;
+    if (!options.pollInterval) options.pollInterval = this.pollInterval;
+
+    const job = this.createJob(options);
+    try {
+      await job.open();
+      await job.uploadData(options.input);
```

Comment:
these might be good places for loggers to write things

Reply:
added logs to each job method instead of here. Couldn't think of any specifics to put here that wouldn't apply to used methods.